torch/csrc/jit/passes/automatic_mixed_precision.cpp

```diff
@@ -0,0 +1,263 @@
+
+#include <torch/csrc/jit/passes/automatic_mixed_precision.h>
+
+#include <c10/core/ScalarType.h>
+#include <c10/util/Exception.h>
+#include <c10/util/Optional.h>
+#include <torch/csrc/jit/ir/ir.h>
+#include <torch/csrc/jit/jit_log.h>
+#include <torch/csrc/jit/passes/quantization/helper.h>
+
+#include <stack>
+#include <unordered_set>
+
+namespace torch {
+namespace jit {
+
+namespace {
+
+struct AutocastScope {
+  Value* instance = nullptr;
+  bool enabled = false;
+};
+
+// If we have an autocast instance, return it
+//
+// This is the pattern we're looking for (this is done after
+//  autocast.__init__() has been inlined)
+//
+// %4 : bool = prim::Constant[value=1]()
+// %5 : __torch__.torch.cuda.amp.autocast_mode.autocast = prim::CreateObject()
+//  = prim::SetAttr[name="_enabled"](%5, %4)
+//
+// Notes:
+//  1. There's no guarantee that the autocast instance is in the same block
+//    as the prim::Enter() node
+//  2. `prim::SetAttr` must follow `prim::CreateObject()` in the same block,
+//    but there might be other nodes in between
+//
+c10::optional<AutocastScope> parseAutocast(Value* value) {
+  if (value->node()->kind() == prim::CreateObject) {
+    const auto class_name = getModuleName(value);
+    if (class_name &&
+        *class_name == "__torch__.torch.cuda.amp.autocast_mode.autocast") {
+      // Search for `prim::SetAttr[name="_enabled"]`
+      for (Use use : value->uses()) {
+        if (use.user->kind() == prim::SetAttr &&
+            use.user->s(attr::name) == "_enabled") {
+          const auto def = use.user->input(1)->node();
+          if (def->kind() != prim::Constant) {
+            AT_ERROR("Autocast argument must be a constant");
```

Comment:
Is this user facing or internal to torchscript?

Reply:
This would be an user-facing limitation and error. One of the todo items is to make the error a bit more user friendly, but the limitation is fundamental in how TorchScript works.

Reply:
Okay yeah, was just thinking if there is some guidance on why this would happen would be good to add to the message.